### DISCUSSION TITLE
Converting JSON schemas to syntax diagrams

### DISCUSSION THREAD

**Discussion by Markus-Elfring** [json-schema-org/community#201] [Q&A]
:thought_balloon: I would like to convert selected JSON schemas to [syntax diagrams](https://en.wikipedia.org/wiki/Syntax_diagram "Description for the usage of syntax diagrams").
Which tools can help any more with corresponding data visualisations? :thinking:

**Comment by jviotti** [comment]
I know @awwright is working on mapping JSON Schema definitions to automata. I would presume that getting a syntax diagram out of an automata would be a much simpler transformation.

**Comment by awwright** [comment]
This ought to be fully possible, and I'm building towards doing exactly this. It just requires a bit of work. Unfortunately there's not very much tooling around ABNF; it seems that developers typically re-implement the ABNF by hand. So I'm building that out.

However I'm not sure how comprehensible the diagrams would be. Here is the ABNF representing the set of `{ "multipleOf": 3 }`:

```
JSON-text-div3 = ws value-div3 ws

value-div3 = false / null / true / object / array / number-div3 / string

number-div3 = "0" / number-div3-positive

number-div3-positive =  (([(("3" / "6" / "9") *("0" / "3" / "6" / "9"))] ("2" / "5" / "8") *(("0" / "3" / "6" / "9")
    / ("1" / "4" / "7") *("0" / "3" / "6" / "9") ("2" / "5" / "8")) ("1" / "4" / "7") / ("3" / "6" / "9")) /
    [(("3" / "6" / "9") *("0" / "3" / "6" / "9"))] (("1" / "4" / "7") / ("2" / "5" / "8") *(("0" / "3" / "6" / "9") /
    ("1" / "4" / "7") *("0" / "3" / "6" / "9") ("2" / "5" / "8")) (("1" / "4" / "7") *("0" / "3" / "6" / "9")
    ("1" / "4" / "7") / ("2" / "5" / "8"))) *((("1" / "4" / "7") / ("2" / "5" / "8") *("0" / "3" / "6" / "9")
    ("2" / "5" / "8")) *(("0" / "3" / "6" / "9") / ("1" / "4" / "7") *("0" / "3" / "6" / "9") ("2" / "5" / "8"))
    (("1" / "4" / "7") *("0" / "3" / "6" / "9") ("1" / "4" / "7") / ("2" / "5" / "8")) / ("0" / "3" / "6" / "9")
    / ("2" / "5" / "8") *("0" / "3" / "6" / "9") ("1" / "4" / "7")) ((("1" / "4" / "7") / ("2" / "5" / "8")
    *("0" / "3" / "6" / "9") ("2" / "5" / "8")) *(("0" / "3" / "6" / "9") / ("1" / "4" / "7") *("0" / "3" / "6" / "9")
    ("2" / "5" / "8")) ("1" / "4" / "7") / ("2" / "5" / "8"))) *("0" / "3" / "6" / "9")

```

Most JSON schema keywords can be expressed as grammar like this. It's just a matter of how complicated it is. `{ "multipleOf": 7 }` is hundreds of times longer than this.

For the curious: multipleOf is one of these compatible keywords because of two things. First, you can append a `0` to a number and it does not change the factors, it adds a `10` to the list of factors. Adding the `multipleOf` also does not change if that `multipleOf` is in the list of factors. Hence, the number ends with `*("0" / "3" / "6" / "9")` — appending a 0, 3, 6, or 9 to any number that is a multiple of 3 will remain a multiple of 3. Then the first part of the expression is building up every single number that's a multiple of 3 that cannot be built from this rule (e.g. 27 cannot be made of only 0, 3, 6, and 9s).

In short, I think JSON Schema _is_ the human-friendly format — and ABNF (or regular expression) is the way of describing how a computer would parse a JSON document meeting those specifications.

Thanks for the ping @jviotti

**Reply by awwright** [reply]
@Markus-Elfring Can you tell me a little bit more about what you're looking to do?

**Reply by Markus-Elfring** [reply]
I am developing JSON schema files for some use cases.
I need to collaborate with software users who are not so familiar with corresponding data formats.
:thought_balloon: Thus I would like to present evolving data structures also by the means of syntax diagrams.

**Reply by awwright** [reply]
Are these users implementing their own parser? Usually it's much easier to use an existing parser, unless you're doing something really fancy.

**Reply by Markus-Elfring** [reply]
:thought_balloon: Syntax diagrams should generally help with the communication and understanding for various data formats, shouldn't they?

Special software components will usually be constructed for customised (JSON) data structures also with [the help of parser generators](https://json-schema.org/implementations.html#generators-from-schemas "Software generation"), won't they? :thinking:

**Comment by plc-dev** [comment]
I was also looking for a tool that transforms a given JSON-schema into syntax diagrams.

I have since come across some libraries that can transform subsets of JSON-schemas into a constrained form of BNFs ([GBNF](https://github.com/ggerganov/llama.cpp/blob/master/grammars/README.md)) [1].
The use case is very different (constrained LLM generation), but since naturally syntax/railroad diagram generation from BNFs is readily available, it should be relatively easy to transform GBNF to a flavor supported by syntax diagram generators (typically W3C?), such as [this](https://rr.red-dove.com/ui).

[1]:
- [json-schema-to-gbnf](https://github.com/adrienbrault/json-schema-to-gbnf)
- [GBNFGen](https://github.com/intrinsiclabsai/gbnfgen)
- [LocalAI PKG](https://github.com/mudler/LocalAI/blob/v1.40.0/pkg/grammar/json_schema.go)

**Reply by plc-dev** [reply]
I think the conclusion by @awwright, that JSON-schema _is_ the human-readable format, may be correct.
The automatic translation of even a very simple JSON-schema yields a very verbose BNF and thus also a very verbose railroad diagram.

Take the following extremely simple JSON-schema:
```json
{
  "properties": {
    "simpleString": {
      "type": "string"
    }
  }
}
```
which yields the following GNBF (using [JSON-Schema-to-GBNF](https://adrienbrault.github.io/json-schema-to-gbnf/)):
```ebnf
root-simpleString ::= "\"simpleString\"" ":" ws01 string
root ::= undefined ws01


value  ::= (object | array | string | number | boolean | null) ws

object ::=
  "{" ws (
    string ":" ws value
    ("," ws string ":" ws value)*
  )? "}"

array  ::=
  "[" ws01 (
            value
    ("," ws01 value)*
  )? "]"

string ::=
  "\"" (string-char)* "\""

string-char ::= [^"\\] | "\\" (["\\/bfnrt] | "u" [0-9a-fA-F] [0-9a-fA-F] [0-9a-fA-F] [0-9a-fA-F]) # escapes

number ::= integer ("." [0-9]+)? ([eE] [-+]? [0-9]+)?
integer ::= "-"? ([0-9] | [1-9] [0-9]*)
boolean ::= "true" | "false"
null ::= "null"

# Optional space: by convention, applied in this grammar after literal chars when allowed
ws ::= ([ \t\n] ws)?
ws01 ::= ([ \t\n])?
```

which requires some corrections to get the W3C EBNF notation

```ebnf
root ::= "{" ws01 root-simpleString "}" 
root-simpleString ::= "\"simpleString"\"

value  ::= (object | array | string | number | boolean | null)

object ::=
  "{" ws (
    string ":" ws value
    ("," ws string ":" ws value)*
  )? "}"

array  ::=
  "[" ws01 (
            value
    ("," ws01 value)*
  )? "]"

string ::= "\" (string-char)* "\"

string-char ::= [^"\\] | "\\" (["\\/bfnrt] | "u" [0-9a-fA-F] [0-9a-fA-F] [0-9a-fA-F] [0-9a-fA-F])

number ::= integer ("." [0-9]+)? ([eE] [-+]? [0-9]+)?
integer ::= "-"? ([0-9] | [1-9] [0-9]*)
boolean ::= "true" | "false"
null ::= "null"

ws ::= ([ \t\n] ws)?
ws01 ::= ([ \t\n])?
```

and finally produces this railroad diagram (using [Railroad Diagram Generator](https://rr.red-dove.com/ui)):
![Screenshot 2025-03-15 141330](https://github.com/user-attachments/assets/fc04364c-1502-4567-b27a-797ab62fa006)

For my purposes (mostly documentation of DSL's that are derived from Typescript types), one of the following seems more appropriate:
- the manual creation of a more concise EBNF (or directly syntax diagrams)
- the use of another diagram notation (e.g. [UML](https://github.com/SOM-Research/jsonSchema-to-uml))

**Reply by aakashkolli** [reply]
@plc-dev I saw you mentioned finding libraries like [JSON-Schema-to-GBNF](https://adrienbrault.github.io/json-schema-to-gbnf/) that convert JSON schemas to GBNF grammars. I'm exploring this same path, for the use case of constrained LLM generation. I'm running into the issue where the output is verbose and needs manual corrections to become valid EBNF. 

I was wondering if you came across any tools that can generate a EBNF directly from a JSON schema, as opposed to a GBNF grammar? Or have you found that manually creating an EBNF grammar is the most effective way to get a concise result?
